File: client/game_card/Assets/Scripts/App/Common/LogicMgr/LogicBase.cs
using UnityEngine;
using UnityEditor;

namespace Utopia
{
    public abstract class LogicBase : Utopia.EventMgr
    {
        protected TimerProxy m_timerProxy = Core.ins.timer.CreateTimerProxy();
        protected LogicMgr m_logicMgr = null;

        public LogicBase()
        {
        }

        public void SetLogicMgr(LogicMgr owner)
        {
            UnityEngine.Debug.Assert(null == m_logicMgr);
            UnityEngine.Debug.Assert(null != owner);
            m_logicMgr = owner;
        }

        public abstract EAppLogicName GetModuleName();
        protected abstract void OnInit();
        protected abstract void OnStart();
        protected abstract void OnUpdate();
        protected abstract void OnRelease();

        public void Init()
        {
            this.OnInit();
        }

        public void Start()
        {
            this.OnStart();
        }

        public void Update()
        {
            this.OnUpdate();
        }

        public void Release()
        {
            this.OnRelease();
            this.ClearAll();
            m_timerProxy.ClearAll();
        }
    }
}

File: client/game_card/LuaScript/app/logic/logic_base.lua

---@class LogicBase:EventMgr
LogicBase = LogicBase or class("LogicBase", EventMgr)

function LogicBase:ctor(logic_mgr, name)
    LogicBase.super.ctor(self)
    self._name = name
    ---@type LogicMgr
    self._logic_mgr = logic_mgr
    ---@type LuaApp
    self._app = self._logic_mgr._app
    ---@type EventBinder
    self._event_binder = EventBinder:new()
end

function LogicBase:get_name()
    return self._name
end

function LogicBase:init()
    self:_on_init()
end

function LogicBase:start()
    self:_on_start()
end

function LogicBase:stop()
    self:_on_stop()
end

function LogicBase:release()
    self:cancel_all()
    self._event_binder:release_all()
    self:_on_release()
end

function LogicBase:_on_init()

end

function LogicBase:_on_start()

end

function LogicBase:_on_stop()

end

function LogicBase:_on_release()

end

